web-xdg-open.1
.TH wxo 1 web-xdg-open
.SH NAME
web-xdg-open - open links in the program configured for xdg-open instead of a
normal browser

.SH SYNOPSIS
.B web-xdg-open
.I LINK

.SH DESCRIPTION
.B web-xdg-open
tries to open links with the
.B default filetype
handler. For example images are opened in the default image viewer as
configured by
.B xdg-utils.

.SH CONFIGURATION

You can configure
.B web-xdg-open
by overriding it's default behaviour for a file group. This can be done in one
of:

.in 10
.B $XDG_CONFIG_HOME/wxorc

.in 10
.B ~/.config/wxorc

.in 7
The file groups are:
.in 10
- image
.in 10
- video
.in 10
- text
.in 10
- git
.in 10
- gif
.in 10
- pdf
.in 10
- clipboard
.in 7
Additionally you can also override a
.I other
function to provide handling of other patterns that
.B web-xdg-open
does not already recognise. This function receives a second parameter containing the same link
stripped of the any query string it might have.

.in 7
For example, to open text files in neovim one can define this function in the
config file:

.in 10
text() {
    dl "$1" >"$TEMPFILE" && $TERMINAL -e bash -c "nvim '$TEMPFILE'"
.br
}

.in 7
.B IMPORTANT NOTE:
.br
For
.B web-xdg-open
to work properly it is very important that exit statuses are
propagated outside these functions. To do this either chain all commands with
.I &&
or append
.I ||return
to the command.

If you don't do this then the fallback might not be used and
.B web-xdg-open
will be very frustrating to use.

.B Example:
.in 10
video() {
    command1 && command2
.br
}
.br
.I OR
.br
video() {
    command1 || return
.br
    command2 || return
.br
}

.in 7
When defining these functions there are a few utilities that
.B web-xdg-open
defines you can make use of when making your config file. You can also
override them but this is highly discouraged.

.TP
.B notify()
A simple wrapper around
.B notify-send(1)
that never returns an error exit status.

.TP
.B dl()
Downloads the passed link and outputs the downloaded data to
.B stdout
returning an error exit status if the download fails for any reason and uses
.B notify()
to warn about it.

.TP
.B fallback()
Is used in case
.B web-xdg-open
fails for any reason. By default it opens the link in
.I $BROWSER
.br
Change this variable in the config file (or globally) to set your default web
browser.

.TP
.B $TEMPFILE
A variable that is guaranteed to hold the name of a valid temporary file in the
format
.I tmp.XXXXXXXXXX.EXT
where
.I XXXXXXXXXX
are random characters and
.I EXT
is the file extension. This file is also guaranteed to be
.I deleted
after
.B web-xdg-open
exits.

.SH FAQ
.TP
.B What is 'clipboard' used for?
The 'clipboard' function will copy the file's contents to the clipboard.
Currently only files ending in
.I *.pub
are copied to the clipboard, as these are usually public rsa keys.

.TP
.B How do I set web-xdg-open as my default browser?
xdg-mime default web-xdg-open.desktop x-scheme-handler/http
.br
xdg-mime default web-xdg-open.desktop x-scheme-handler/https

.TP
.B How can I have a really cool confy config?
That's easy! Just download
.I https://github.com/mendess/spell-book/blob/master/runes/wxorc
😎.

.SH AUTHORS
Written by Pedro Mendes <pedro.mendes.26@gmail.com> originally in 2020.

.SH LICENSE
GPLv3

.SH SEE ALSO
.BR notify-send(1)
.BR mpv(1)
.BR xclip(1)
.BR xdg-open(1)
.BR xdg-mime(1)
.BR youtube-dl(1)


web-xdg-open.sh
#!/bin/bash
set -o pipefail

[[ ! "$1" ]] &&
    echo "Usage: $0 link" &&
    exit 1

readonly CONFIG="${XDG_CONFIG_HOME:-$HOME/.config}/wxorc"
#shellcheck disable=2155
readonly APP_NAME="$(basename "$0")"

## ============== PUBLIC OVERRIDABLE METHODS ================
notify() {
    notify-send -a "$APP_NAME" "$@"
    true
}

image() {
    dl "$1" >"$TEMPFILE" && xdg-open "$TEMPFILE"
}

video() {
    if command -V mpv &>/dev/null; then
        notify 'Playing in mpv' "$1" -u low
        mpv "$1" || {
            notify 'Retrying in mpv' "$1" -u low
            mpv "$1"
        }
    else
        notify 'Downloading' "$1"
        if ! error=$(youtube-dl --no-color "$1" --output "$TEMPFILE"); then
            notify 'Download failed' "$error" -u critical
            return 1
        else
            xdg-open "$TEMPFILE"
        fi
    fi
}

text() {
    dl "$1" >"$TEMPFILE" && xdg-open "$TEMPFILE"
}

matematica() {
    text "$@"
}

git() {
    dl "$1" >"$TEMPFILE" && xdg-open "$TEMPFILE"
}

gif() {
    if command -V mpv &>/dev/null; then
        mpv --loop-file "$1"
    else
        dl "$1" && xdg-open "$TEMPFILE"
    fi
}

pdf() {
    dl "$1" >"$TEMPFILE" && xdg-open "$TEMPFILE"
}

clipboard() {
    if command -V xclip &>/dev/null; then
        dl "$1" | xclip && notify 'Key coppied to clipboard' -u low
    else
        notify 'xclip not installed' \
            'Install it or override the behaviour of clipboard' \
            -u low
    fi
}

archive() {
    false
    #dl "$1" >"$TEMPFILE" && xdg-open "$TEMPFILE"
}

fallback() {
    notify 'Using fallback' -u low
    if [[ "$BROWSER" ]]; then
        "$BROWSER" "$1" &
        disown
    else
        notify 'Browser not set' \
            "Please set the \$BROWSER enviroment variable" \
            -u critical
        false
    fi
    exit
}

dl() {
    __bash_not_smart_enough() {
        local error
        if ! error=$(curl -# --fail "$1" 3>&2 2>&1 1>&3); then
            notify 'Download failed' \
                "$(tail -1 <<<"$error" | sed -E 's/curl:\s*\([0-9]*\)//g')" \
                -u critical
            return 1
        fi
    }
    __bash_not_smart_enough "$@" 2>&1
}

other() {
    false
}

__temp_file() {
    TEMPFILE=$(mktemp) || return 1
    if [[ "$1" ]]; then
        name="$(basename "$1")"
        [[ "$name" = */* ]] && return 0
        mv "$TEMPFILE" "$TEMPFILE-$name" || return 1
        TEMPFILE="$TEMPFILE-$name"
    fi
}

## ============== PRIVATE METHODS ================

__cleanup() {
    sleep 30
    [[ "$TEMPFILE" ]] && rm -f "$TEMPFILE"
    :
}

trap __cleanup EXIT

readonly target="$1"

__try() {
    clean_link=${target,,}
    clean_link=${clean_link%\?*}
    case "$clean_link" in
        *github*blob*)
            raw_link="${target//github/raw.githubusercontent}"
            raw_link="${raw_link//blob\//}"
            git "$raw_link"
            ;;
        *.txt | *.sh | *.c | *.cpp | *.h | *.hpp | *.rs | *.spell | *.tool | *.hs | *.conf | *.tex | *.log | *.lhs | *.java | *.py | *.awk | *.pl | *.cs | *.css | *.js | *.scss | *.lua | *.bash | *makefile | *.md | *.yml | *.yaml)
            text "$target"
            ;;
        *.nb)
            matematica "$target"
            ;;
        *.png | *.jpg | *.jpeg | *.bmp | *.tiff | *.svg | *.ico | *.img | *.webp)
            image "$target"
            ;;
        *.gif)
            gif "$target"
            ;;
        *.pdf | *.djvu)
            pdf "$target"
            ;;
        *.tar | *.bz2 | *.rar | *.Z | *.7z | *.gz | *.xz | *.zip)
            archive "$target"
            ;;
        *youtube.com* | *youtu.be* | *vimeo* | *clips.twitch.tv* | *soundcloud.com* | https://v.redd.it* | https://t.co* | *medal.tv* | *streamable.com* | *.avi | *.webm | *.mp4 | *.mp3 | *.wav | *.flac | *.midi | *.dvdrip | *.cam | *.mkv | *.mov | *.mpeg | *.flv | *.mpg | *.mp2 | *.mpv | *.m4p | *.m4v | *.wmv | *.qt | *.swf | *.avchd | *.m4a | *.ogg | *.wma | *.amv | *.mpa | *.ra | *.rax | *.raw | *.smf | *.snd | *.sng | *.swa | *.hma | *.aac | *.ac3 | *.eac3 | *.Vorbis | *.pcm)
            video "$target"
            ;;
        *.pub)
            clipboard "$target"
            ;;
        *) other "$target" "$clean_link" ;;
    esac
}

#shellcheck disable=1090
. "$CONFIG"

__temp_file "$1" || fallback "$target"

__try || fallback "$target"
